### ISSUE TITLE
Odd/bad widget behavior in a second window

### ISSUE THREAD

**Issue by emailman** [lawsie/guizero#130]
These widgets don't work right in a second window.  Their bad behavior seems consistent with tkinter.

CheckBox reports only 0, not 1
TextBox string is not read
PushButton does not show text and is not correct size
Combo does not list the options until the bar is clicked

`
from guizero import *

def check():
    print("check box:",cbX.value)
    print("text box:", tbY.value)
    print("combo option:", combo1.value)

window1 = App(title="Window 1")
window2 = App(title="Window 2")

cbX = CheckBox(window2, text="X", command=check)
tbY = TextBox(window2, grid=[0, 1])
pbZ = PushButton(window2, text="Click Here", command=check)
combo1 = Combo(window2, options=["one", "two", "three"])

window1.display()
`

**Comment by Coal0** [comment]
I was afraid this was gonna come up at some point.
Unfortunately, this issue is closely tied to the fundamental philosophy of this project: ease of use.
Tkinter only supports one `tk.Tk` instance at runtime (it is tied to a global Tcl interpreter). To create new windows, you're supposed to instantiate a `tk.Toplevel` object:

```python
import tkinter as tk

root = tk.Tk()
window_2 = tk.Toplevel(root)
```

But `App.__init__` creates unconditionally creates a new `tk.Tk` object.

---

So, what now?
There's a couple of things we could do:

1. Change the documentation to mention that only one `App` object may be alive at a given time. In this case, we should probably change the implementation too, and print an error message in case an attempt is made to create a second instance. Perhaps not feasible, because there is an apparent need for secondary windows.

2. Change the implementation of `App` to create either a `tk.Tk` instance or a `tk.Toplevel` instance, depending on whether or not a root window already exists. Added complexity: lots of error handling.

3. Add a `Window` type with much of the same functionality (through mixins) as `App`.

@martinohanlon  @lawsie  What do you say?

**Comment by emailman** [comment]
I appreciate the philosophy of guizero is simplicity.  Supporting multiple windows would be a "nice to have", not  a "must have".

**Comment by Coal0** [comment]
I think it would fit in nicely in the [0.5 'New features' release](https://github.com/lawsie/guizero/projects/3). Just in case, here's a quick fix:

```python
import guizero as gz
import guizero.mixins as mix
import guizero.tkmixins as tkmix


class Window(
    mix.ContainerMixin,
    tkmix.ScheduleMixin, 
    tkmix.DestroyMixin, 
    tkmix.FocusMixin, 
    tkmix.ReprMixin,
    tkmix.SizeMixin
):
    def __init__(
                 self,
                 master,
                 title="guizero",
                 width=500,
                 height=500,
                 layout="auto",
                 bg=None
                ):
        self.tk = tk.Toplevel(master=master)

        self.description = "[Window] object"
        self.tk.title(str(title))
        self.tk.geometry(str(width)+"x"+str(height))
        self._layout_manager = layout

        if bg is not None:
            self.tk.configure(background=str(bg))

        self.tk.update()

    @property
    def title(self):
        return self.tk.title()

    @title.setter
    def title(self, text):
        self.tk.title(str(text))

    def display(self):
        self.tk.mainloop()

    def on_close(self, command):
        self.tk.wm_protocol("WM_DELETE_WINDOW", command)

```

Untested - it probably has rough edges and glaring bugs, sorry.

Credit where due, most code taken ad verbum from ['App.py'](https://github.com/lawsie/guizero/blob/dev/guizero/App.py)

**Comment by martinohanlon** [comment]
ill have to do some investigation and mull over what is the 'right' thing to do about multi window guis.

i dont think it will be a short term aim to add support though.

**Comment by emailman** [comment]
Hi Coal.  Thanks for the quick fix.  Here's my quick test.  Everything works as expected.
`
import tkinter as tk
import guizero.mixins as mix
import guizero.tkmixins as tkmix
from guizero import *


class Window(
    mix.ContainerMixin,
    tkmix.ScheduleMixin,
    tkmix.DestroyMixin,
    tkmix.FocusMixin,
    tkmix.ReprMixin,
    tkmix.SizeMixin
):
    def __init__(
                 self,
                 master,
                 title="guizero",
                 width=500,
                 height=500,
                 layout="auto",
                 bg=None
                ):
        self.tk = tk.Toplevel(master=master)

        self.description = "[Window] object"
        self.tk.title(str(title))
        self.tk.geometry(str(width)+"x"+str(height))
        self._layout_manager = layout

        if bg is not None:
            self.tk.configure(background=str(bg))

        self.tk.update()

    @property
    def title(self):
        return self.tk.title()

    @title.setter
    def title(self, text):
        self.tk.title(str(text))

    def display(self):
        self.tk.mainloop()

    def on_close(self, command):
        self.tk.wm_protocol("WM_DELETE_WINDOW", command)


def pb1():
    print("text box 1:", tbx1.value)


def pb2():
    print("text box 2:", tbx2.value)


def cb1():
    print("check box 1:", cbx1.value)


def cb2():
    print("check box 2:", cbx2.value)


def combo1(x):
    print("Combo 1:", x)


def combo2(x):
    print("Combo 2:", x)


root = tk.Tk()
root.withdraw()

window1 = Window(master=None, title="Window 1")
window2 = Window(master=None, title="Window 2")

tbx1 = TextBox(window1)
tbx2 = TextBox(window2)

PushButton(window1, text="Read TextBox 1", command=pb1)
PushButton(window2, text="Read TextBox 2", command=pb2)

cbx1 = CheckBox(window1, text="check box 1", command=cb1)
cbx2 = CheckBox(window2, text="check box 2", command=cb2)

Combo(window1, options=["Red", "Green", "Blue"], command=combo1)
Combo(window2, options=["White", "Black", "Gray"], command=combo2)

PushButton(window1, text="QUIT", command=quit)
PushButton(window2, text="QUIT", command=quit)

window1.display()
`

**Comment by Coal0** [comment]
N.b. it is preferable to create a single `App` instance and pass it as 'master' argument to secondary windows:

```python
import guizero as gz

class Window( ... ):
     ...

app = gz.App(title="Window 1")
window2 = Window(master=app, title="Window 2")
# window3 = Window(master=app, title="Window 3") etc.
```

It's also a bit cleaner than having a hidden root window at runtime.

**Comment by emailman** [comment]
Daniel, noted.  Thanks for providing a cleaner way to handle the secondary windows.

**Comment by emailman** [comment]
This code:
`
app = gz.App(title="Window 1")
window2 = Window(master=app, title="Window 2")
`
Results in this:
Traceback (most recent call last):
  File "C:/Users/eric2/PycharmProjects/minh1/bugfix1.py", line 77, in <module>
    window2 = Window(master=app, title="Window 2")
  File "C:/Users/eric2/PycharmProjects/minh1/bugfix1.py", line 25, in __init__
    self.tk = tk.Toplevel(master=master)
  File "C:\Program Files (x86)\Python36-32\lib\tkinter\__init__.py", line 2336, in __init__
    BaseWidget.__init__(self, master, 'toplevel', cnf, {}, extra)
  File "C:\Program Files (x86)\Python36-32\lib\tkinter\__init__.py", line 2286, in __init__
    BaseWidget._setup(self, master, cnf)
  File "C:\Program Files (x86)\Python36-32\lib\tkinter\__init__.py", line 2263, in _setup
    if master._last_child_ids is None:
AttributeError: 'App' object has no attribute '_last_child_ids'

Process finished with exit code 1

------------------------------------------------------------------------------------------------------------
This code runs OK:
`
app = gz.App(title="Window 1")
window2 = Window(master=None, title="Window 2")
`

**Comment by emailman** [comment]
This code runs without errors:

app = gz.App(title="Window 1")
window2 = Window(master=None, title="Window 2")

----------------------------
But, this code results in the exception shown below:

app = gz.App(title="Window 1")
window2 = Window(master=app, title="Window 2")

---------------------------------------------
Traceback (most recent call last):
  File "C:/Users/eric2/PycharmProjects/minh1/bugfix1.py", line 77, in <module>
    window2 = Window(master=app, title="Window 2")
  File "C:/Users/eric2/PycharmProjects/minh1/bugfix1.py", line 25, in __init__
    self.tk = tk.Toplevel(master=master)
  File "C:\Program Files (x86)\Python36-32\lib\tkinter\__init__.py", line 2336, in __init__
    BaseWidget.__init__(self, master, 'toplevel', cnf, {}, extra)
  File "C:\Program Files (x86)\Python36-32\lib\tkinter\__init__.py", line 2286, in __init__
    BaseWidget._setup(self, master, cnf)
  File "C:\Program Files (x86)\Python36-32\lib\tkinter\__init__.py", line 2263, in _setup
    if master._last_child_ids is None:
AttributeError: 'App' object has no attribute '_last_child_ids'

Process finished with exit code 1

**Comment by Coal0** [comment]
Whoops, my bad. Try this:

```python
app = gz.App(title="Window 1")
window2 = Window(master=app.tk, title="Window 2")
```

Or for a more permanent solution:

```python
class Window( ... ):
    def __init__(self, master, ... ):
        self.tk = tk.Toplevel(master=master.tk)
        ...
```

**Comment by martinohanlon** [comment]
I have created issue #155 "Multiple Windows" and proposed an API to support this.

Let me know what you think.

**Comment by martinohanlon** [comment]
Ok, so I have resolved the issue with launching 'multiple' apps. This isn't the final solution for dealing with multiple windows, but now if you launch multiple `App` objects from the same python program it works, you can share data between them.

Its in a separate branch at the moment but will go into v0.5.0 
https://github.com/lawsie/guizero/commit/32d779ef9d25b4fa121e615750e05ec66d757037

**Comment by martinohanlon** [comment]
resolved in v0.5.0

**Comment by emailman** [comment]
Martin,

It's great to see this implemented.  Thanks for getting it into the new
release.
Eric Mailman

---- Original Message ----
